File: 104.MaximumDepthOfBinaryTree/depth.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque
class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
            
        queue1 = deque([])
        queue2 = deque([])

        queue1.append(root)
        queue2.append(1)

        currDepth = 0
        while queue1 and queue2:
            node = queue1.popleft()
            currDepth = queue2.popleft()
            if node.left:
                queue1.append(node.left)
                queue2.append(currDepth + 1)
            if node.right:
                queue1.append(node.right)
                queue2.append(currDepth + 1)

        return currDepth

File: 104.MaximumDepthOfBinaryTree/Depth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Deque<TreeNode> queue1 = new ArrayDeque<TreeNode>();
        Deque<Integer> queue2 = new ArrayDeque<Integer>();

        queue1.add(root);
        queue2.add(1);

        int currDepth = 0;
        while (!queue1.isEmpty() && !queue2.isEmpty()) {
            TreeNode node = queue1.poll();
            currDepth = queue2.poll();

            if (node.left != null) {
                queue1.add(node.left);
                queue2.add(currDepth + 1);
            }

            if (node.right != null) {
                queue1.add(node.right);
                queue2.add(currDepth + 1);
            }
        }

        return currDepth;
    }

    public itn maxDepth2(TreeNode root) {
        return root == null ? 0 : 1 + Math.max(maxDepth2(root.left), maxDepth2(root.right));
    }
}
